# The Topics of the *nlp4twitter* Course

We will adopt a data science approach to this course in order to find which NLP techniques to study. We will "process" all articles in the [ACM DL](http://dl.acm.org/) (expanded to all records) with the word `twitter` **in the title** (we had to stop somewhere. We could have searched for `twitter` everywhere, ~2.2K results. We did, however, extend the search to the whole database, not only the ACM Full Text DB). 

Specifically, we ran the following query:
```
Search Run Date: 2016-04-15 at 12:46:23 PM EDT

Query Syntax:
"query": { acmdlTitle:(twitter) } 

"filter": 
{owners.owner=GUIDE}
```
which resulted in 1,307 articles matching it (as of the date above, could be more by now!).

Now, ACM DL, as far as I know, doesn't allow you to download > ~1000, I did a trick: I sorted by number of citations *and* provided two ranges: the first was from 2006-2014 (998 records), and the second was from 2015-2016 (309 records), for both `.csv` and `bib` files, which I will use later.

This produced the following 4 files:

In [16]:
acm_files = !ls data/bib/*n.*
acm_files

['data/bib/acm_twitter_2006-2014_citation.bib',
 'data/bib/acm_twitter_2006-2014_citation.csv',
 'data/bib/acm_twitter_2015-2016_citation.bib',
 'data/bib/acm_twitter_2015-2016_citation.csv']

Notice that *each file* is sorted by number of citations independently, but split on publication data. This means that the concatenation of them will not give you a complete citation-sorted file. For now, we are only be interested in the `csv` files, because they're easier to deal with in Notebooks.

In [5]:
import pandas as pd

In [175]:
acm_2006_2014 = pd.read_csv(
    "data/bib/acm_twitter_2006-2014_citation.csv")
acm_2015_2016 = pd.read_csv(
    "data/bib/acm_twitter_2015-2016_citation.csv")

Let's first join the two data frames:

In [176]:
dfs = [acm_2006_2014, acm_2015_2016]
acm = pd.concat(dfs)

But we don't want anything that is not a paper (e.g., we don't want books, or PhD thesis) or really care about anything but the titles (maybe), the years, and particularly, the keywords. We also get rid of nulls, which pushes us to even less records, but fine.

In [177]:
acm = acm[acm['type'].str.contains("article")]
acm = acm[['keywords','title','year']]
acm = acm[acm['keywords'].notnull()]
acm = acm[acm['title'].notnull()]
len(acm)

905

The number 905 is still a gargantuan number of papers to analyze! The objective here is to get a sense of the methods used in those papers. But first, let's get rid of some of the crud. Here we split all keywords into "phrases" using "," as the splitter and convert the Pandas `Series` object into a list of lists. We then convert that list of lists into a single list whose phrases are stripped of formatting and the space character and lower-cased, and finally we get rid of duplicates:

In [182]:
pd.set_option('display.max_colwidth', -1)
acm.head()

keywords  \
0  Twitter, degree of separation, homophily, influential, information diffusion, online social network, pagerank, reciprocity, retweet   
1  Twitter, earthquake, event detection, location estimation, social sensor                                                              
2  microblogging, social media, social network analysis, user intent                                                                     
3  influential, pagerank, twitter                                                                                                        
4  communication networks, diffusion, influence, twitter, word of mouth marketing                                                        

                                                                          title  \
0  What is Twitter, a Social Network or a News Media?                             
1  Earthquake Shakes Twitter Users: Real-time Event Detection by Social Sensors   
2  Why We Twitter: Understanding Microblogging Usage and Communities              
3  TwitterRank: Finding Topic-sensitive Influential Twitterers                    
4  Everyone's an Influencer: Quantifying Influence on Twitter                     

   year  
0  2010  
1  2010  
2  2007  
3  2010  
4  2011

In [184]:
acm.to_csv('data/bib/acm.csv')

In [169]:
keywords = acm['keywords'].str.split(',').tolist()
keywords = [item.lower().strip() 
            for sublist in keywords for item in sublist]
keywords=set(keywords)
len(keywords)

1951

The above process leaves "only" 1951 phrases that are unique keywords to those papers!

In [161]:
keywords

{'opinion leadership',
 'predictive models',
 'similarity',
 'firm innovativeness',
 'identity-based groups',
 'entropy',
 'twitter topic detection',
 'micro-blog',
 'spatial point pattern',
 'german elections',
 'semi-supervised learning',
 'tweet',
 'dialogic loop',
 'web service',
 'name disambiguation',
 'ergm',
 'tobacco use',
 'audience responses',
 'regrets',
 'command and control',
 'process mining',
 'risk',
 'media selection',
 'internet',
 'behavior modeling',
 'social behavioral biometrics',
 'document clustering',
 'affect',
 'news media',
 'linguistic analysis',
 'tagging',
 'communication networks',
 'election campaign',
 'graph-based',
 'loneliness',
 'latent factor models',
 'cmc',
 'browsers',
 'social influence',
 'data journalism',
 'virtual identity',
 'unsupervised learning',
 'quantitative evaluation',
 'welfare',
 'social networking potential',
 'impact indicator',
 'public shaming',
 'recency ranking',
 'protest mobilization',
 'sentiment and emotions classific